service/src/main/java/org/cbioportal/service/impl/GenesetCorrelationServiceImpl.java

```diff
@@ -0,0 +1,264 @@
+        	
+        	double correlationValue = 0;
+        	// arrays need to be at least 2 long to calculate correlation: 
+        	if (geneAndGenesetValues.geneValues.length > 2) {
+	    		// calculate spearman correlation
+	        	correlationValue = spearmansCorrelation.correlation(geneAndGenesetValues.geneValues, geneAndGenesetValues.genesetValues);
+        	}
+    		// filter out the ones below correlationThreshold
+        	if (Math.abs(correlationValue) < correlationThreshold) {
+        		continue;
+        	}
+        	GenesetCorrelation genesetCorrelationItem = new GenesetCorrelation();
+        	genesetCorrelationItem.setEntrezGeneId(entrezGeneId);
+        	genesetCorrelationItem.setHugoGeneSymbol(gene.getHugoGeneSymbol());
+        	genesetCorrelationItem.setCorrelationValue(correlationValue);
+        	genesetCorrelationItem.setExpressionGeneticProfileId(expressionProfile.getStableId());
+        	genesetCorrelationItem.setzScoreGeneticProfileId(zscoresProfile.getStableId());
+        	result.add(genesetCorrelationItem);
+        }        
+		// return sorted
+        sortResult(result);
+		return result;
+	}
+
+
+	private GeneticProfile getLinkedZscoreProfile(GeneticProfile expressionProfile) {
+		
+		//Find the related z-score profile via the genetic_profile_link table:
+    	List<GeneticProfile> referringProfiles = geneticProfileService.getGeneticProfilesReferringTo(expressionProfile.getStableId());
+    	GeneticProfile zscoresProfile = null;
+    	for (GeneticProfile referringProfile : referringProfiles) {
+    		//use the first z-score profile we can find in this list of referring profiles (normally there should be only 1 anyway):
+    		if (referringProfile.getDatatype().equals(DataType.Z_SCORE)) {
+    			zscoresProfile = referringProfile;
+    			break;
+    		}
+    	}
+    	//if none found, give clear error message...something is wrong with this study:
+        if (zscoresProfile == null) {
+        	throw new IllegalArgumentException("The expression profile [" + expressionProfile.getStableId() + "] linked to the given "
+					+ "gene set scores profile does not have a corresponding z-scores profile in this study.");
+        }
+        return zscoresProfile;
+	}
+
+	private void sortResult(List<GenesetCorrelation> result) {
+        Comparator<GenesetCorrelation> comparator = new Comparator<GenesetCorrelation>() {
+            @Override
+            public int compare(GenesetCorrelation o1, GenesetCorrelation o2) {
+            	
+        		//descending order, also check for NaN
+        		if (o1.getCorrelationValue().isNaN())
+        			return 1;
+        		if (o2.getCorrelationValue().isNaN())
+        			return -1;
+        		
+        		if (o1.getCorrelationValue() < o2.getCorrelationValue())
+        			return 1;
+        		if (o1.getCorrelationValue() > o2.getCorrelationValue())
+        			return -1;
+        		
+        		return 0;
+            }
+        };
+		Collections.sort(result, comparator);		
```

Comment:
Can you use lambda expression here using `result.sort` method, instead of Comparator?